scope.go

```diff
@@ -750,3 +1073,25 @@ func (s *histogramSnapshot) Values() map[float64]int64 {
 func (s *histogramSnapshot) Durations() map[time.Duration]int64 {
 	return s.durations
 }
+
+type scopeExpiryTracker struct {
+	expiredCounters   map[*counter]struct{}
+	expiredGauges     map[*gauge]struct{}
+	expiredHistograms map[*histogram]struct{}
+
+	numCounters   int
```

Comment:
why do we keep this, why not just do `len(expiredCounters)` in the `numMetrics` function, not sure it gives any performance win, since length is fast to access on a slice/map anyway. it's more likely to get out of date if we cache it separately

Reply:
We're tracking all counters, not just the expired ones with numCounters. We only expire the scope when all the metrics have been expired, so on that loop, len of expiredCounters will be nothing. I also need to keep track in a separate var as we read the len in a RLock.